ADR-214: PointerEvents (#214)
decentraland/adr: content/ADR-214-pointer-events-sdk-component.md

  optional float analog = 6;
}

enum InputAction {
  IA_POINTER = 0;
  IA_PRIMARY = 1;
  IA_SECONDARY = 2;
  IA_ANY = 3;
  IA_FORWARD = 4;
  IA_BACKWARD = 5;
  IA_RIGHT = 6;
  IA_LEFT = 7;
  IA_JUMP = 8;
  IA_WALK = 9;
  IA_ACTION_3 = 10;
  IA_ACTION_4 = 11;
  IA_ACTION_5 = 12;
  IA_ACTION_6 = 13;
}

// PointerEventType is a kind of interaction that can be detected.
enum PointerEventType {
  PET_UP = 0;
  PET_DOWN = 1;
  PET_HOVER_ENTER = 2;
  PET_HOVER_LEAVE = 3;
}

```

## Semantics

## Determine the interactable entities

A component called `PointerEvents` is designed to signal the Renderer which entities are interactable. It is also used to provide information about "call to action placeholders" which are recommended to appear as visual cues in the screen. These cues are specialized per input action, enabling the renderer to select a special texture for each action key-mapping.

Based on the `PointerEvents` component, at the "executeRaycast" stage of the tick ([ADR-148](/adr/ADR-148)), a continuous raycast ([ADR-200](/adr/ADR-200)) is added to process the events to be sent to the scene.

### Mesh selection for PointerEvent raycast

> The eligible meshes for raycast are the ones that met any of this criteria:

Reviewer: ```suggestion
The eligible meshes for raycast are the ones that meet any of this criteria:
```

Any entity with `UiTransform` ([ADR-124](/adr/ADR-124)) can be eligible for pointer events if they have a `PointerEvents` component, all other UI entities will not be eligible and pointer events will be bubbled up until finding an entity that matches the criteria. A rectangular mesh occupying the whole area designated by the `UiTransform` MUST be used.

In UI entities, contrary to the render order ([ADR-151](/adr/ADR-151)), the events MUST bubble up from the leaves of the tree up to the root until finding an entity that matches the pointer event. In that case, the propagation of the event MUST stop, and the event MUST be added to the `PointerEventsResult`.

## Input commands

For the ECS to work, "events" and "state" must be separated into two different categories. A challenge is faced here, since "events" or "interrupts" are not compatible with data representations at one moment in time (like the state of the ECS). To overcome this, all the input events of the frame must be queued and processed at the "executeRaycast" stage of the tick for all the scenes that are running. Naturally, this compute MUST count towards the quota/limit suggested by ADR-148.

> Then, in the "executeRaycast" step of the scene tick (on the renderer side), that queue will be consumed and "InputCommands" will be produced. An InputCommand is a record in a PointerEventResult component listing all the pertinent input actions for that entity. It is done to process when the pointer is "hovering" an entity, when it leaves, when the key down and key up in input. All of that can happen in the same frame. For example, a cube can be aimed at, clicked, and released the click on the same frame.

Reviewer: ```suggestion
Then, in the "executeRaycast" step of the scene tick (on the renderer side), that queue will be consumed and "InputCommands" will be produced. An InputCommand is a record in a PointerEventResult component listing all the pertinent input actions for that entity. This is done to process when the pointer is "hovering" an entity, when it leaves, when the key down and key up in input. All of that can happen in the same frame. For example, a cube can be aimed at, clicked, and released the click on the same frame.
```

This design enables rapid games with low input lag and without losing information regardless of the framerate of the scene and renderer.

> It also enables high resolution events for cases of moving entities, e.g. assuming one is pointing still, with one finger on the trigger, if the engine puts an entity in front of someone for one frame, the scene will receive the "Hover Enter" event for that unique frame, and if one is fast enough, it may also be clicked.

Reviewer: ```suggestion
It also enables high resolution events for cases of moving entities, e.g. assuming one is pointing still, with one finger on the trigger, if the engine puts an entity in front of someone for one frame, the scene will receive the "HoverEnter" event for that unique frame, and if one is fast enough, it may also be clicked.
```
